numba/core/cgutils.py

```diff
@@ -1013,56 +1014,59 @@ def cbranch_or_continue(builder, cond, bbtrue):
     """
     Emit a memcpy to the builder.
 
-    Copies each element of dst to src. Unlike the C equivalent, each element
-    can be any LLVM type.
+    Source and destination types can be any type, but the types must be the
+    same.
 
-    Assumes
-    -------
-    * dst.type == src.type
-    * count is positive
+    Call either as:
+
+    - memcpy(builder, dst, src, len), or
+    - memcpy(builder, dst, src, count, itemsize)
     """
-    # Note this does seem to be optimized as a raw memcpy() by LLVM
-    # whenever possible...
-    assert dst.type == src.type
-    with for_range(builder, count, intp=count.type) as loop:
-        out_ptr = builder.gep(dst, [loop.index])
-        in_ptr = builder.gep(src, [loop.index])
-        builder.store(builder.load(in_ptr), out_ptr)
+    if dst.type != src.type:
+        msg = f'memcpy requires the same types; got {dst.type} and {src.type}'
+        raise TypingError(msg)
 
+    _raw_mem_intrinsic(builder, 'llvm.memcpy', dst, src, count, itemsize)
 
-def _raw_memcpy(builder, func_name, dst, src, count, itemsize, align):
-    size_t = count.type
-    if isinstance(itemsize, int):
-        itemsize = ir.Constant(size_t, itemsize)
 
-    memcpy = builder.module.declare_intrinsic(func_name,
-                                              [voidptr_t, voidptr_t, size_t])
-    is_volatile = false_bit
-    builder.call(memcpy, [builder.bitcast(dst, voidptr_t),
-                          builder.bitcast(src, voidptr_t),
-                          builder.mul(count, itemsize),
-                          is_volatile])
+def memmove(builder, dst, src, count, itemsize=1):
+    """
+    Emit a memmove to the builder.
 
+    Source and destination types can be any type, but the types must be the
+    same.
 
-def raw_memcpy(builder, dst, src, count, itemsize, align=1):
-    """
-    Emit a raw memcpy() call for `count` items of size `itemsize`
-    from `src` to `dest`.
+    Call either as:
+
+    - memmove(builder, dst, src, len), or
+    - memmove(builder, dst, src, count, itemsize)
     """
-    return _raw_memcpy(builder, 'llvm.memcpy', dst, src, count, itemsize, align)
+    _raw_mem_intrinsic(builder, 'llvm.memmove', dst, src, count, itemsize)
 
 
-def raw_memmove(builder, dst, src, count, itemsize, align=1):
-    """
-    Emit a raw memmove() call for `count` items of size `itemsize`
-    from `src` to `dest`.
-    """
-    return _raw_memcpy(builder, 'llvm.memmove', dst, src, count,
-                       itemsize, align)
+def _raw_mem_intrinsic(builder, func_name, dst, src, count, itemsize):
+    if isinstance(count, int):
+        count = ir.Constant(intp_t, count)
+
+    if isinstance(itemsize, int):
+        itemsize = ir.Constant(intp_t, itemsize)
+
+    if isinstance(itemsize, ir.Constant) and itemsize.constant == 1:
+        length = count
+    else:
+        length = builder.mul(count, itemsize)
+
+    memcpy = builder.module.declare_intrinsic(func_name,
```

Comment:
```suggestion
    func_impl = builder.module.declare_intrinsic(func_name,
```
(and refactor, this interface is intended to be generic for `memcpy` and `memmove`)?

Reply:
Will refactor to make this generic.